### PR TITLE
Use sdist to find files to monitor, fixes #3019 for Simple Cabal files

### LINKED ISSUE #3019: Last step in cabal-nix sometimes takes a loooooooooong time
Building `hackage-server`:

```
cabal-nix new-build -v3
...
link: linkables are ...
LinkableM (2016-01-02 08:32:56 UTC) main:Main
   [DotO /Users/e/wt/projects/tuf/hackage-server/dist-newstyle/build/hackage-server-0.5.1/build/PackageTests/PackageTests-tmp/Main.o]
LinkableM (2016-01-02 08:32:55 UTC) main:Distribution.Server.Packages.Unpack
   [DotO /Users/e/wt/projects/tuf/hackage-server/dist-newstyle/build/hackage-server-0.5.1/build/PackageTests/PackageTests-tmp/Distribution/Server/Packages/Unpack.o]
LinkableM (2016-01-02 08:32:56 UTC) main:Distribution.Server.Packages.UnpackTest
   [DotO /Users/e/wt/projects/tuf/hackage-server/dist-newstyle/build/hackage-server-0.5.1/build/PackageTests/PackageTests-tmp/Distribution/Server/Packages/UnpackTest.o]
LinkableM (2015-12-30 15:01:50 UTC) main:Distribution.Server.Util.Parse
   [DotO /Users/e/wt/projects/tuf/hackage-server/dist-newstyle/build/hackage-server-0.5.1/build/PackageTests/PackageTests-tmp/Distribution/Server/Util/Parse.o]
LinkableM (2015-12-30 15:01:49 UTC) main:Distribution.Server.Util.GZip
   [DotO /Users/e/wt/projects/tuf/hackage-server/dist-newstyle/build/hackage-server-0.5.1/build/PackageTests/PackageTests-tmp/Distribution/Server/Util/GZip.o]
/Users/e/wt/projects/tuf/hackage-server/dist-newstyle/build/hackage-server-0.5.1/build/PackageTests/PackageTests is up to date, linking not required.
*** Deleting temp files:
Deleting: 
*** Deleting temp dirs:
Deleting: /var/folders/94/vkm1c1l17cddyvzk2252hnp40000gs/T/ghc81573_0
```

and then.. nothing.. The `cabal-nix` process is using some 25% CPU at this point, and I have no idea what it's doing. That temp directory mentioned on the last line is already gone, so it's not just deleting. What it is doing is anyone's guess.. I CTRL-C'ed it after 12 minutes. Retried, same thing. Wiped `dist-newbuild`, same thing. I've seen long times before but never something like this. No idea what's happening. Will reboot machine and see if that resolved anything. (Incidentally, I thought maybe it was busy waiting for the `hackage-server` executable, but the server was not running at this point.)

### PR DESCRIPTION
**Pull request by ezyang**
First two commits are just simple updates. Third commit is a refactoring I needed to reuse sdist in cabal-install. Last commit is the fix.

I'm reasonably confident this should work, but I haven't tested quite enough (there aren't any integration tests for new-build boohoo).

### CODE REVIEW

**Comment by ezyang** [comment]
CC @dcoutts

File: Cabal/Distribution/Simple/Utils.hs [hunk 1]
```diff
@@ -663,8 +668,20 @@ xargs maxSize rawSystemFun fixedArgs bigArgs =
 findFile :: [FilePath]    -- ^search locations
          -> FilePath      -- ^File Name
          -> IO FilePath
-findFile searchPath fileName =
-  findFirstFile id
+findFile = findFile0 id
+
+findDirFile :: FilePath   -- ^ working directory (not included in result)
```
Comment (23Skidoo) [inline Cabal/Distribution/Simple/Utils.hs]: Please add Haddock comments to all new top-level definitions in this file.

File: Cabal/Distribution/Simple/SrcDist.hs [hunk 2]
```diff
@@ -130,37 +131,46 @@ listPackageSources :: Verbosity          -- ^ verbosity
                    -> [PPSuffixHandler]  -- ^ extra preprocessors (include
                                          -- suffixes)
                    -> IO ([FilePath], [FilePath])
-listPackageSources verbosity pkg_descr0 pps = do
+listPackageSources = listPackageSources' "."
+
+listPackageSources' :: FilePath           -- ^ current working directory
```
Comment (23Skidoo) [inline Cabal/Distribution/Simple/SrcDist.hs]: Please add a Haddock comment.

**Comment by 23Skidoo** [comment]
The `.cabal` file can sometimes contain references to non-existing files (via `extra-source-files`, for example), so may be worth adding a test for that.

**Comment by ezyang** [comment]
Well, tests for new-build in general would be good...

IIUC, sdist will just not report files that don't exist. This can tickle a known bug which I mentioned at https://github.com/haskell/cabal/issues/3019#issuecomment-204284682 I don't know what happens in this situation.

**Comment by 23Skidoo** [comment]
Also, the sandbox code uses the setup script to list sources (added in c7aa58148fe881932be8dbf4092ce0672377afd3/#1305). The reason for that is apparently #403.

**Comment by ezyang** [comment]
Oh that's lovely! I thought I was going to have to implement that but I guess not! (Note that #403 only applies for Custom setup scripts, so we can avoid an exec when the build type is simple.)

**Comment by 23Skidoo** [comment]
> so we can avoid an exec when the build type is simple.

`setupWrapper` does that automatically for you.

**Comment by ezyang** [comment]
Oh, I see that `internalSetupMethod` uses `inDir` to change the working directory, which means it won't work in parallel. (Oh! That explains why we force an external process for parallel builds.)

**Comment by ezyang** [comment]
OK, so an alternate way to implement this is to use SetupWrapper, and ditch the CWD refactoring. The current patchset only has a modest benefit, which is that it can be done same process even when we are multithreaded (not true for setup wrapper.) @23Skidoo, what do you think?

**Comment by 23Skidoo** [comment]
I like the current solution a bit better, since there are less moving parts, so unless there's a compelling reason to use `setupWrapper` (is #403 a concern?), I'd prefer to keep it.

**Comment by ezyang** [comment]
We must use setupWrapper for Custom Cabal scripts, so I'm going to have to implement that code path anyhow.

**Comment by 23Skidoo** [comment]
Then it's probably easier to just use `setupWrapper` uniformly.

**Comment by ezyang** [comment]
This branch now also contains some of @grayjay's fixes, which needed to be merged, were very helpful and should be backported.

**Comment by cartazio** [comment]
question: at least as of a few days ago with when i built the nix-local-builds branch, the output for `cabal help new-build` was the standard(ish) docs for `cabal help install`. Is that supposed to be, or is the cabal help output semi stale?

**Comment by ezyang** [comment]
The docs are not up to date at the moment. (although, I think the flag descriptions are accurate)

**Comment by ezyang** [comment]
Is there an easy way to kick the AppVeyor build here?

**Comment by 23Skidoo** [comment]
Restarted the build.

**Comment by ezyang** [comment]
Merged in.

**Pull request closed without merging**